traffic_ops/app/lib/API/Configs/ApacheTrafficServer.pm

```diff
@@ -0,0 +1,1959 @@
+	elsif ( $fname eq "plugin.config" )           { $scope = 'profile' }
+	elsif ( $fname eq "storage.config" )          { $scope = 'profile' }
+	elsif ( $fname eq "sysctl.conf" )             { $scope = 'profile' }
+	elsif ( $fname =~ /url_sig_.*\.config/ )      { $scope = 'profile' }
+	elsif ( $fname eq "volume.config" )           { $scope = 'profile' }
+	elsif ( $fname eq "bg_fetch.config" )         { $scope = 'cdn' }
+	elsif ( $fname =~ /cacheurl.*\.config/ )      { $scope = 'cdn' }
+	elsif ( $fname =~ /hdr_rw_.*\.config/ )       { $scope = 'cdn' }
+	elsif ( $fname =~ /regex_remap_.*\.config/ )  { $scope = 'cdn' }
+	elsif ( $fname eq "regex_revalidate.config" ) { $scope = 'cdn' }
+	elsif ( $fname =~ /set_dscp_.*\.config/ )     { $scope = 'cdn' }
+	elsif ( $fname eq "ssl_multicert.config" )    { $scope = 'cdn' }
+	else {
+		$scope = $self->db->resultset('Parameter')->search( { -and => [ name => 'scope', config_file => $fname ] } )->get_column('value')->first();
+		if ( !defined($scope) ) {
+			$self->app->log->error("Filename not found.  Setting Server scope.");
+			$scope = 'server';
+		}
+	}
+
+	return $scope;
+}
+
+#takes the server name or ID and turns it into a server object that can reference either, making either work for the request.
+sub server_data {
+	my $self = shift;
+	my $id   = shift;
+
+	my $server_obj;
+
+	#if an ID is passed, look up by ID.  Otherwise, look up by hostname.
+	if ( $id =~ /^\d+$/ ) {
+		$server_obj = $self->db->resultset('Server')->search( { id => $id } )->single;
+	}
+	else {
+		$server_obj = $self->db->resultset('Server')->search( { host_name => $id } )->single;
+	}
+
+	return $server_obj;
+}
+
+#takes the profile name or ID and turns it into a server object that can reference either, making either work for the request.
+sub profile_data {
+	my $self = shift;
+	my $id   = shift;
+
+	#if an ID is passed, look up by ID.  Otherwise, look up by profile name.
+	my $profile_obj;
+	if ( $id =~ /^\d+$/ ) {
+		$profile_obj = $self->db->resultset('Profile')->search( { id => $id } )->single;
+	}
+	else {
+		$profile_obj = $self->db->resultset('Profile')->search( { name => $id } )->single;
+	}
+
+	return $profile_obj;
+}
+
+#takes the server name or ID and turns it into a server object that can reference either, making either work for the request.
+sub cdn_data {
+	my $self = shift;
+	my $id   = shift;
+	my $cdn_obj;
+
+	if ( $id =~ /^\d+$/ ) {
+		$cdn_obj = $self->db->resultset('Cdn')->search( { id => $id } )->single;
+	}
+	else {
+		$cdn_obj = $self->db->resultset('Cdn')->search( { name => $id } )->single;
+	}
+
+	return $cdn_obj;
+}
+
+#generates the comment at the top of config files.
+sub header_comment {
+	my $self = shift;
+	my $name = shift;
+
+	my $text = "# DO NOT EDIT - Generated for " . $name . " by " . &name_version_string($self) . " on " . `date`;
+	return $text;
+}
+
+#retrieves parameter data for a specific server by searching by the server's assigned profile.
+sub param_data {
```

Comment:
the only difference i see between `sub param_data` and `sub profile_param_data` is the following piece:

if ( $value =~ /^STRING __HOSTNAME__$/ ) {
     $value = "STRING " . $server_obj->host_name . "." . $server_obj->domain_name;
}

is that bit of code only relevant when pulling a "take and bake" config file at the server level?

any way you can just collapse the 2 methods?